Wrap system clock for modifying in tests and undo

frank-weinberg opened the pull request:
instead of manipulating the internals of DefaultClockModel

frank-weinberg committed:
Wrap system clock for modifying in tests and undo

instead of manipulating the internals of DefaultClockModel

brian-brazil commented on src/com/carolinarollergirls/scoreboard/utils/ScoreBoardClock.java:
```diff
@@ -0,0 +1,66 @@
+package com.carolinarollergirls.scoreboard.utils;
+
+import com.carolinarollergirls.scoreboard.defaults.DefaultScoreBoardEventProvider;
```
I don't think we should have a circular dependency between packages like this, this class should go in defaults.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/utils/ScoreBoardClock.java:
```diff
@@ -0,0 +1,66 @@
+	public String getProviderId() { return ""; }
+
+	public static ScoreBoardClock getInstance() {
+		if (instance == null) {
```
If you're doing this you need locking.

A static initialiser might be simpler.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/defaults/DefaultClockModel.java:
```diff
@@ -386,15 +388,16 @@ private DefaultClockSnapshotModel(ClockModel clock) {
 			} catch ( Exception e ) { }
-			startSystemTime = System.currentTimeMillis();
+			startSystemTime = scoreBoardClock.getCurrentTime();
+			scoreBoardClock.addScoreBoardListener(this);
```
Currently all listens are via the DefaultScoreBoardModel, so this should work the same way for consistency.

frank-weinberg committed:
Merge branch 'dev' into scoreBoardClock

brian-brazil commented on src/com/carolinarollergirls/scoreboard/utils/ScoreBoardClock.java:
```diff
@@ -0,0 +1,66 @@
+		updateTime();
+		stopCounter++;
+	}
+	public synchronized void start(Boolean doCatchUp) {
```
`boolean`

brian-brazil reviewed (commented):
Using events for this seems a bit weird, I think this logic would better live either in DefaultClockModel or have all this pass via ScoreBoardModel.

frank-weinberg committed:
address review comments

Initiate timer ticks in ScoreBoardClock instead of UpdateClockTimerTask
di not use events in ScoreBoardClock

frank-weinberg commented:
Using events felt weird when writing the code. Your comments pushed me towards using a different approach.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/utils/ScoreBoardClock.java:
```diff
@@ -0,0 +1,88 @@
+	
+	private List<ScoreBoardClockClient> clients = new ArrayList<ScoreBoardClockClient>();
+	
+	public static final long CLOCK_UPDATE_INTERVAL = 200; /* in ms */
```
Why not set this to 50 instead of dividing?

There's also an override of this value in the android code that should be deleted.

frank-weinberg replied:
UpdateClockTimerTask uses the unmodified value as tick interval.
(everything else should be addressed in the new commit)

brian-brazil commented on src/com/carolinarollergirls/scoreboard/utils/ScoreBoardClock.java:
```diff
@@ -0,0 +1,88 @@
+		stopCounter--;
+	}
+	
+	public void registerClient(ScoreBoardClockClient client) {
```
This should be synchronized

brian-brazil commented on src/com/carolinarollergirls/scoreboard/utils/ScoreBoardClock.java:
```diff
@@ -0,0 +1,88 @@
+	public static final long CLOCK_UPDATE_INTERVAL = 200; /* in ms */
+
+	public interface ScoreBoardClockClient {
+		void updateTime(long ms);
```
A comment here mentioning this is passed the current time rather than an offset would be useful.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/utils/ScoreBoardClock.java:
```diff
@@ -0,0 +1,88 @@
+	private int stopCounter = 0;
+	private long lastRewind = 0;
+	
+	private static Timer timer = new Timer();
```
Why is this static?

brian-brazil reviewed (commented):
This looks better.

frank-weinberg committed:
review comments

frank-weinberg committed:
addendum to previous commit
